### PR TITLE
Test discovery with adding/removing capabilties

### PR DESCRIPTION
**Pull request by stevemar**

### CODE REVIEW

**Review by rhagarty — APPROVED** [review]
Looks good

File: ?
Comment (markstur) [inline ?:1]: mutable default arguments are bad

Comment (markstur) [inline ?:1]: here too

**Review by markstur — CHANGES_REQUESTED** [review]
see inline

File: ?
Comment (stevemar) [reply ?:1]: this isn't in a sig

Comment (stevemar) [reply ?:1]: ok, will fix

**Review by stevemar — COMMENTED** [review]
lgtm

**Pull request merged**